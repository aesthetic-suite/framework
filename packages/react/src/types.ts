// Any must be used here for consumers to be typed correctly.
/* eslint-disable @typescript-eslint/no-explicit-any */

import React from 'react';
import { Theme, ClassNameSheetVariants } from '@aesthetic/core';
import { ClassName, Direction, ThemeName } from '@aesthetic/types';

export type ClassNameGenerator<T> = (
  ...keys: (undefined | null | false | T | ClassNameSheetVariants)[]
) => ClassName;

// CONTEXT

export type DirectionContextType = Direction;

export interface DirectionProviderProps {
  children: NonNullable<React.ReactNode>;
  /** Explicit direction to use. */
  direction?: Exclude<Direction, 'neutral'>;
  /** Render an inline element instead of block. */
  inline?: boolean;
  /** Locale aware string to deduce the direction from. */
  value?: string;
}

export type ThemeContextType = Theme;

export interface ThemeProviderProps {
  children: NonNullable<React.ReactNode>;
  /** Theme to activate. */
  name?: ThemeName;
}

// HOCs

export interface WrapperComponent {
  WrappedComponent: React.ComponentType<any>;
}

export interface WrapperProps {
  /** Gain a reference to the wrapped component. */
  wrappedRef?: React.Ref<any>;
}

export interface WrappedProps {
  /** The ref passed by the `wrappedRef` prop. */
  ref?: React.Ref<any>;
}

export interface WithDirectionWrappedProps extends WrappedProps {
  /** The current direction. Provided by `withDirection`. */
  direction: Direction;
}

export interface WithThemeWrappedProps extends WrappedProps {
  /** The theme object. Provided by `withTheme`. */
  theme: Theme;
}

export interface WithStylesWrappedProps<T = string> extends WrappedProps {
  /** Function to generate CSS class names from a style sheet. Provided by `withStyles`. */
  cx: ClassNameGenerator<T>;
}
