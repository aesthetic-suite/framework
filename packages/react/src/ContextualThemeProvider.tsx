import React, { useState, useEffect } from 'react';
import { getTheme, renderThemeStyles } from '@aesthetic/core';
import ThemeContext from './ThemeContext';
import useDirection from './useDirection';
import { ThemeProviderProps } from './types';

/**
 * Rendered nested within the page to provide a contextual theme.
 */
export default function ContextualThemeProvider({ children, name }: Required<ThemeProviderProps>) {
  const [className, setClassName] = useState('');
  const direction = useDirection();
  const theme = getTheme(name);

  // Render styles when theme/direction change
  useEffect(() => {
    setClassName(renderThemeStyles(theme, { direction }));
  }, [theme, direction]);

  return (
    <ThemeContext.Provider value={theme}>
      <div className={className} data-theme={name}>
        {children}
      </div>
    </ThemeContext.Provider>
  );
}
