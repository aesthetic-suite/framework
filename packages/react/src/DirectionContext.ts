import React from 'react';
import { getActiveDirection } from '@aesthetic/core';
import { DirectionContextType } from './types';

export default React.createContext<DirectionContextType>(getActiveDirection() || 'ltr');
