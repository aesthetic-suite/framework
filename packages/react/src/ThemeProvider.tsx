import React, { useState, useEffect, useCallback } from 'react';
import { changeTheme, getActiveTheme, getTheme, subscribe, unsubscribe } from '@aesthetic/core';
import ThemeContext from './ThemeContext';
import { ThemeProviderProps } from './types';

/**
 * Rendered at the root to provide the theme to the entire application.
 */
export default function ThemeProvider({ children, name = '' }: ThemeProviderProps) {
  const [themeName, setThemeName] = useState(name);
  const theme = themeName ? getTheme(themeName) : getActiveTheme();

  // Listen to theme changes that occur outside of the provider
  const handleChangeTheme = useCallback(
    (nextTheme) => {
      if (nextTheme !== themeName) {
        setThemeName(nextTheme);
      }
    },
    [themeName],
  );

  useEffect(() => {
    subscribe('change:direction', handleChangeTheme);
    subscribe('change:theme', handleChangeTheme);

    return () => {
      unsubscribe('change:direction', handleChangeTheme);
      unsubscribe('change:theme', handleChangeTheme);
    };
  }, [handleChangeTheme]);

  // Update state when the `name` prop changes
  useEffect(() => {
    if (name) {
      changeTheme(name, false);
      setThemeName(name);
    }
  }, [name]);

  return <ThemeContext.Provider value={theme}>{children}</ThemeContext.Provider>;
}
