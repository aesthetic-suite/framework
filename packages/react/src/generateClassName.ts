import { ClassNameSheet } from '@aesthetic/core';
import { ClassName } from '@aesthetic/types';
import { arrayLoop, isObject, objectLoop } from '@aesthetic/utils';

export default function generateClassName<T extends unknown[]>(
  keys: T,
  classNames: ClassNameSheet<string>,
): ClassName {
  const list: string[] = [];

  arrayLoop(keys, (key) => {
    // Variants
    if (isObject(key)) {
      objectLoop(key, (value, type) => {
        if (!value || !type) {
          return;
        }

        const variant = `${type}_${value}`;

        objectLoop(classNames, (clx) => {
          if (clx?.variants?.[variant]) {
            list.push(clx.variants[variant]);
          }
        });
      });

      // Rules/styles
    } else if (typeof key === 'string' && classNames[key]?.class) {
      list.push(classNames[key]!.class);
    }
  });

  return list.join(' ');
}
