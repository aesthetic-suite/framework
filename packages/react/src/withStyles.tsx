import React from 'react';
import hoistNonReactStatics from 'hoist-non-react-statics';
import { LocalSheet } from '@aesthetic/core';
import useStyles from './useStyles';
import { WithStylesWrappedProps, WrapperProps, WrapperComponent } from './types';

/**
 * Wrap a React component with an HOC that injects the style to class name transfer function.
 */
export default function withStyles<T = unknown>(sheet: LocalSheet<T>) /* infer */ {
  return function withStylesComposer<Props extends object = {}>(
    WrappedComponent: React.ComponentType<Props & WithStylesWrappedProps<keyof T>>,
  ): React.NamedExoticComponent<Omit<Props, keyof WithStylesWrappedProps> & WrapperProps> &
    WrapperComponent {
    const WithStyles = React.memo(function WithStyles({
      wrappedRef,
      ...props
    }: Props & WrapperProps) {
      const cx = useStyles(sheet);

      return <WrappedComponent {...(props as any)} ref={wrappedRef} cx={cx} />;
    }) as ReturnType<typeof withStylesComposer>;

    hoistNonReactStatics(WithStyles, WrappedComponent);

    WithStyles.displayName = `withStyles(${WrappedComponent.displayName || WrappedComponent.name})`;

    WithStyles.WrappedComponent = WrappedComponent;

    return WithStyles;
  };
}
