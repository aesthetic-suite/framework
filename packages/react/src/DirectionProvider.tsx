import React from 'react';
import getDirection from 'direction';
import { Direction } from '@aesthetic/types';
import DirectionContext from './DirectionContext';
import { DirectionProviderProps } from './types';

/**
 * Explicitly define a direction or automatically infer a direction based on a string of text.
 * Will render an element with a `dir` attribute set.
 */
export default function DirectionProvider({
  children,
  direction,
  inline,
  value,
}: DirectionProviderProps) {
  const Tag = inline ? 'span' : 'div';
  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
  const dir: Direction = direction || getDirection(value) || 'ltr';

  return (
    <DirectionContext.Provider value={dir}>
      <Tag dir={dir}>{children}</Tag>
    </DirectionContext.Provider>
  );
}
