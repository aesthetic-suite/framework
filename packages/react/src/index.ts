/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import DirectionContext from './DirectionContext';
import DirectionProvider from './DirectionProvider';
import ThemeContext from './ThemeContext';
import ThemeProvider from './ThemeProvider';
import ContextualThemeProvider from './ContextualThemeProvider';
import useDirection from './useDirection';
import useStyles from './useStyles';
import useTheme from './useTheme';
import withDirection from './withDirection';
import withStyles from './withStyles';
import withTheme from './withTheme';

export * from '@aesthetic/core';
export * from './types';

export {
  DirectionContext,
  DirectionProvider,
  ThemeContext,
  ThemeProvider,
  ContextualThemeProvider,
  useDirection,
  useStyles,
  useTheme,
  withDirection,
  withStyles,
  withTheme,
};
