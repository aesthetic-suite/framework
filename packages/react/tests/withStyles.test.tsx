/* eslint-disable react/jsx-no-literals */

import React from 'react';
import { render } from 'rut-dom';
import { aesthetic } from '@aesthetic/core';
import {
  setupAesthetic,
  teardownAesthetic,
  getRenderedStyles,
  darkTheme,
  lightTheme,
} from '@aesthetic/core/lib/testing';
import withStyles from '../src/withStyles';
import ThemeProvider from '../src/ThemeProvider';
import { WithStylesWrappedProps } from '../src/types';
import { createStyleSheet, ButtonProps, Wrapper } from './__mocks__/Button';

describe('withStyles()', () => {
  beforeEach(() => {
    setupAesthetic(aesthetic);
  });

  afterEach(() => {
    teardownAesthetic(aesthetic);
  });

  function BaseButton({
    children,
    cx,
    block,
    disabled,
    large,
    small,
  }: ButtonProps &
    WithStylesWrappedProps<
      'button' | 'button_block' | 'button_disabled' | 'button_large' | 'button_small'
    >) {
    return (
      <button
        type="button"
        className={cx('button', block && 'button_block', disabled && 'button_disabled', {
          // eslint-disable-next-line no-nested-ternary
          size: large ? 'lg' : small ? 'sm' : 'df',
        })}
      >
        {children}
      </button>
    );
  }

  // Props need to exist for generic inheritance to work correctly
  interface BaseComponentProps {
    unknown?: unknown;
  }

  function BaseComponent(props: BaseComponentProps & WithStylesWrappedProps<string>) {
    return null;
  }

  function WrappingComponent({ children }: { children?: React.ReactNode }) {
    return <ThemeProvider name="day">{children || <div />}</ThemeProvider>;
  }

  it('inherits name from component `constructor.name`', () => {
    const Wrapped = withStyles(createStyleSheet())(BaseComponent);

    expect(Wrapped.displayName).toBe('withStyles(BaseComponent)');
  });

  it('inherits name from component `displayName`', () => {
    function DisplayComponent() {
      return null;
    }

    DisplayComponent.displayName = 'CustomDisplayName';

    const Wrapped = withStyles(createStyleSheet())(DisplayComponent);

    expect(Wrapped.displayName).toBe('withStyles(CustomDisplayName)');
  });

  it('stores the original component as a static property', () => {
    const Wrapped = withStyles(createStyleSheet())(BaseComponent);

    expect(Wrapped.WrappedComponent).toBe(BaseComponent);
  });

  it('can bubble up the ref with `wrappedRef`', () => {
    interface RefProps {
      unknown?: boolean;
    }

    // eslint-disable-next-line react/prefer-stateless-function
    class RefComponent extends React.Component<RefProps & WithStylesWrappedProps<string>> {
      render() {
        return <div />;
      }
    }

    let foundRef: Function | null = null;
    const Wrapped = withStyles(createStyleSheet())(RefComponent);

    render<{}>(
      <Wrapped
        wrappedRef={(ref: Function | null) => {
          foundRef = ref;
        }}
      />,
      { wrapper: <WrappingComponent /> },
    );

    expect(foundRef).not.toBeNull();
    expect(foundRef!.constructor.name).toBe('RefComponent');
  });

  it('renders a button and its base styles', () => {
    const Button = withStyles(createStyleSheet())(BaseButton);
    const { root } = render<ButtonProps>(<Button>Child</Button>, {
      wrapper: <Wrapper />,
    });

    expect(root.findOne('button')).toHaveProp(
      'className',
      'p q r s t u v w x y z a1 b1 c1 d1 e1 f1 g1 h1 i1 j1 k1 l1 m1 n1 o1 p1 f g h i j',
    );

    expect(getRenderedStyles('standard')).toMatchSnapshot();
  });

  it('only renders once unless theme or direction change', () => {
    const sheet = createStyleSheet();
    const Button = withStyles(sheet)(BaseButton);
    const spy = jest.spyOn(sheet, 'render');

    const { update } = render<ButtonProps>(<Button sheet={sheet}>Child</Button>, {
      wrapper: <Wrapper />,
    });

    update();
    update();
    update();

    expect(spy).toHaveBeenCalledTimes(2);
  });

  it('re-renders if direction changes', () => {
    const sheet = createStyleSheet();
    const Button = withStyles(sheet)(BaseButton);
    const spy = jest.spyOn(sheet, 'render');

    const { rerender } = render<ButtonProps>(
      <Wrapper>
        <Button sheet={sheet}>Child</Button>
      </Wrapper>,
    );

    expect(spy).toHaveBeenCalledTimes(2);
    expect(spy).toHaveBeenCalledWith(
      expect.anything(),
      darkTheme,
      expect.objectContaining({
        direction: 'ltr',
        theme: 'night',
      }),
    );

    rerender(
      <Wrapper direction="rtl">
        <Button sheet={sheet}>Child</Button>
      </Wrapper>,
    );

    expect(spy).toHaveBeenCalledTimes(3);
    expect(spy).toHaveBeenCalledWith(
      expect.anything(),
      darkTheme,
      expect.objectContaining({
        direction: 'rtl',
        theme: 'night',
      }),
    );
  });

  it('re-renders if theme changes', () => {
    const sheet = createStyleSheet();
    const Button = withStyles(sheet)(BaseButton);
    const spy = jest.spyOn(sheet, 'render');

    const { rerender } = render<ButtonProps>(
      <Wrapper>
        <Button sheet={sheet}>Child</Button>
      </Wrapper>,
    );

    expect(spy).toHaveBeenCalledTimes(2);
    expect(spy).toHaveBeenCalledWith(
      expect.anything(),
      darkTheme,
      expect.objectContaining({
        direction: 'ltr',
        theme: 'night',
      }),
    );

    rerender(
      <Wrapper theme="day">
        <Button sheet={sheet}>Child</Button>
      </Wrapper>,
    );

    expect(spy).toHaveBeenCalledTimes(3);
    expect(spy).toHaveBeenCalledWith(
      expect.anything(),
      lightTheme,
      expect.objectContaining({
        direction: 'ltr',
        theme: 'day',
      }),
    );
  });
});
