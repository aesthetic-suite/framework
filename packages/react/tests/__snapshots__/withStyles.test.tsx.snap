// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withStyles() renders a button and its base styles 1`] = `".a {appearance: none;}.b {background-color: var(--palette-brand-color-40);}.c {border: 0;}.d {cursor: pointer;}.e {display: inline-flex;}.f {font-size: inherit;}.g {margin: 0;}.h {padding: var(--spacing-df) var(--spacing-md);}.i {text-decoration: none;}.j {user-select: auto;}.k {vertical-align: middle;}.l {border-color: var(--palette-neutral-color-30);}.m {border-radius: var(--border-df-radius);}.n {border-style: solid;}.o {border-width: var(--border-df-width);}.p {position: relative;}.q {text-align: center;}.r {color: var(--palette-neutral-color-00);}.s {min-width: 10rem;}.t::-moz-focus-inner {border: 0;}.u::-moz-focus-inner {padding: 0;}.v::-moz-focus-inner {margin: 0;}.w:focus {border-color: var(--palette-primary-bg-focused);}.x:hover {border-color: var(--palette-primary-bg-hovered);}.y:hover {background-color: var(--palette-brand-color-50);}.z:active {border-color: var(--palette-primary-bg-selected);}.a1:active {background-color: var(--palette-brand-color-60);}.b1[disabled] {border-color: var(--palette-primary-bg-disabled);}.c1 {overflow-wrap: normal;}.d1 {word-wrap: normal;}.e1 {word-break: normal;}.f1 {display: block;}.g1 {width: 100%;}.h1 {white-space: normal;}.i1 {overflow: hidden;}.j1 {min-width: 7.50rem;}.k1 {padding: var(--spacing-sm) var(--spacing-df);}.l1 {min-width: 12.50rem;}.m1 {padding: var(--spacing-md) var(--spacing-lg);}.n1 {opacity: 0.5;}.a {appearance: none;}.b {background-color: var(--palette-brand-color-40);}.c {border: 0;}.d {cursor: pointer;}.e {display: inline-flex;}.f {font-size: inherit;}.g {margin: 0;}.h {padding: var(--spacing-df) var(--spacing-md);}.i {text-decoration: none;}.j {user-select: auto;}.k {vertical-align: middle;}.l {border-color: var(--palette-neutral-color-30);}.m {border-radius: var(--border-df-radius);}.n {border-style: solid;}.o {border-width: var(--border-df-width);}.p {position: relative;}.q {text-align: center;}.r {color: var(--palette-neutral-color-00);}.s {min-width: 10rem;}.t::-moz-focus-inner {border: 0;}.u::-moz-focus-inner {padding: 0;}.v::-moz-focus-inner {margin: 0;}.w:focus {border-color: var(--palette-primary-bg-focused);}.x:hover {border-color: var(--palette-primary-bg-hovered);}.y:hover {background-color: var(--palette-brand-color-50);}.z:active {border-color: var(--palette-primary-bg-selected);}.a1:active {background-color: var(--palette-brand-color-60);}.b1[disabled] {border-color: var(--palette-primary-bg-disabled);}.c1 {overflow-wrap: normal;}.d1 {word-wrap: normal;}.e1 {word-break: normal;}.f1 {display: block;}.g1 {width: 100%;}.h1 {white-space: normal;}.i1 {overflow: hidden;}.j1 {min-width: 7.50rem;}.k1 {padding: var(--spacing-sm) var(--spacing-df);}.l1 {min-width: 12.50rem;}.m1 {padding: var(--spacing-md) var(--spacing-lg);}.n1 {opacity: 0.5;}"`;
